src/serial.rs

```diff
@@ -111,6 +116,97 @@ cfg_if! {
     }
 }
 
+/// Types for configuring a serial interface.
+pub mod config {
+    use crate::time::rate::{Baud, Extensions};
+
+    // Reexport stop bit enum from PAC. In case there is a breaking change,
+    // provide a compatible enum from this HAL.
+    pub use crate::pac::usart1::cr2::STOP_A as StopBits;
+
+    /// Parity generation and checking. If odd or even parity is selected, the
+    /// underlying USART will be configured to send/receive the parity bit in
+    /// addtion to the data bits.
+    #[cfg_attr(feature = "defmt", derive(defmt::Format))]
+    #[derive(Clone, Copy, PartialEq)]
+    pub enum Parity {
+        /// No parity bit will be added/checked.
+        ParityNone,
+        /// The MSB transmitted/received will be generated/checked to have a
+        /// even number of bits set.
+        ParityEven,
+        /// The MSB transmitted/received will be generated/checked to have a
+        /// odd number of bits set.
+        ParityOdd,
+    }
+
+    /// Configuration struct for [`Serial`](super::Serial) providing all
+    /// communication-related / parameters.
+    #[derive(Clone, Copy, PartialEq)]
+    pub struct Config {
```

Comment:
Something I've learned recently is using #[non_exhaustive] on structs:

https://rust-lang.github.io/rfcs/2008-non-exhaustive.html#structs-1

```suggestion
    #[derive(Clone, Copy, PartialEq)]
    #[non_exhaustive]
    pub struct Config {
```

With that, it would be easier to add fields later on, with it being backwards
compatible. But with this change, the Config struct could not be configured by
the user anymore like that `Config {...}`.

But I like this behavior more. Because of the fields being public, the user could
still match on the internal fields and read them. It would only require us, to 
implement "setters". But as this is already the case, I think it is fine as is.

But I would document the fact, that it is not constructable by the user and point 
towards the setters instead with an example, something like this:

```rust
/// # Example
/// ```
/// Config::new().baudrate(115_200.Bd()).stopbits(StopBits::STOP1)
/// ```
pub struct Config {
    // ...
}
```

Reply:
That's sounds like a really good idea as we are discussing to extend the config struct one day (for RS-485). I added documentation for the above aproach with `new` and `default`.

I attempted to get the documentation checked with `cargo test` and the best shot I came up so far was building the tests for the host like:
```
$ cargo test --doc --features=stm32f303xc,defmt --target=x86_64-unknown-linux-gnu
```
This helped, sorting my initial issues with them but it does feel clumsy. Do you know a better way to get the examples checked? At least building for the target should be possible but I did not get any errors for a deliberately broken one.